File: mlpack/allknn.cpp
#include <mlpack/core.hpp>
#include "neighbor_search.hpp"
#include "unmap.hpp"

using namespace mlpack;
using namespace mlpack::neighbor;
using namespace mlpack::tree;

int main(int argc, char** argv)
{
    // Contains query points.
    arma::mat query(2,3);
    query << 5 << 6 << 2.75 << arma::endr << 1.45 << 2 << 0.75 << arma::endr;

    // Load the data.
    arma::mat data;
    data::Load("../data/fisheriris_data.csv", data, true);
    // Choose the last two columns.
    data = data.rows(2,3);    
    
    std::vector<size_t> oldFromNewRefs;
    std::vector<size_t> oldFromNewQueries;
    arma::mat distancesOut;
    arma::Mat<size_t> neighborsOut;
    
    // Calculate the all 2-nearest-neighbors .
    BinarySpaceTree<bound::HRectBound<2>, QueryStat<NearestNeighborSort> >
    refTree(data, oldFromNewRefs, 20);
    
    BinarySpaceTree<bound::HRectBound<2>, QueryStat<NearestNeighborSort> >*
    queryTree = new BinarySpaceTree<bound::HRectBound<2>,
    QueryStat<NearestNeighborSort> >(query, oldFromNewQueries, 20);
    
    AllkNN* allknn = new AllkNN(&refTree, queryTree, data, query, false);    
    allknn->Search(2, neighborsOut, distancesOut);
    
    arma::Mat<size_t> neighbors;
    arma::mat distances;
    
    // Map the results back to the correct places.
    Unmap(neighborsOut, distancesOut, oldFromNewRefs, oldFromNewQueries,
          neighbors, distances);
    
    // Clean up.
    delete queryTree;
    delete allknn;
    
    // Show the results.
    Log::cout << distances.t() << std::endl;
    Log::cout << neighbors.t() << std::endl;

    return 0;
}

File: matlab/allknn.m
% Load the Data.
data = csvread('../data/fisheriris_data.csv');
% Choose the last two columns.
data = data(:,3:4);
% Contains query points.
query = [5 1.45;6 2;2.75 .75]

% Calculate the all 2-nearest-neighbors .
[n,d]=knnsearch(data,query,'K',2,'distance','euclidean', 'NSMethod', 'kdtree');

% Show the results.
d
n

% Plot the results
gscatter(data(:,1), data(:,2))
line(query(:,1), query(:,2),'marker','x','color','k',...
   'markersize',10,'linewidth',2,'linestyle','none')
line(data(n,1), data(n,2),'color',[.5 .5 .5],'marker','o',...
   'linestyle','none','markersize',10)